homeassistant/components/camera/push.py

```diff
@@ -0,0 +1,174 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+API_URL = "/api/camera_push/{entity_id}"
+
+CONF_BUFFER_SIZE = "cache"
+
+DEFAULT_NAME = 'Push Camera'
+
+BLANK_IMAGE_SIZE = (640, 480)
+
+ATTR_FILENAME = "filename"
+
+REQUIREMENTS = ['pillow==5.0.0']
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+    vol.Optional(CONF_BUFFER_SIZE, default=1): cv.positive_int,
+    vol.Optional(CONF_TIMEOUT, default=5): cv.positive_int,
+})
+
+
+async def async_setup_platform(hass, config, async_add_devices,
+                               discovery_info=None):
+    """Set up the Push Camera platform."""
+    cameras = [PushCamera(config.get(CONF_NAME),
+                          config.get(CONF_BUFFER_SIZE),
+                          config.get(CONF_TIMEOUT))]
+
+    hass.http.register_view(CameraPushReceiver(cameras))
+
+    async_add_devices(cameras)
+
+
+class CameraPushReceiver(HomeAssistantView):
+    """Handle pushes from remote camera."""
+
+    url = API_URL
+    name = 'api:camera:push'
+
+    def __init__(self, cameras):
+        """Initialize CameraPushReceiver with camera entity."""
+        self._cameras = cameras
+
+    async def post(self, request, entity_id):
+        """Accept the POST from Camera."""
+        try:
+            (_camera,) = [camera for camera in self._cameras
+                          if camera.entity_id == entity_id]
+        except ValueError:
+            return self.json_message('Unknown Push Camera',
+                                     HTTP_BAD_REQUEST)
+
+        try:
+            data = await request.post()
+            _LOGGER.debug("Received Camera push: %s", data['image'])
+            await _camera.update_image(data['image'].file.read(),
+                                       data['image'].filename)
+        except ValueError:
+            return self.json_message('Invalid POST', HTTP_BAD_REQUEST)
+
+
+class PushCamera(Camera):
+    """The representation of a Push camera."""
+
+    def __init__(self, name, buffer_size, timeout):
+        """Initialize push camera component."""
+        super().__init__()
+        self._name = name
+        self._motion_status = False
+        self._last_trip = None
+        self._filename = None
+        self._expired = None
+        self._state = STATE_IDLE
+        self._timeout = datetime.timedelta(seconds=timeout)
+        self.queue = deque([], buffer_size)
+
+        from PIL import Image
+        import io
+
+        image = Image.new('RGB', BLANK_IMAGE_SIZE)
+
+        imgbuf = io.BytesIO()
+        image.save(imgbuf, "JPEG")
+
+        self.queue.append(imgbuf.getvalue())
+        self._current_image = imgbuf.getvalue()
+
+    @property
+    def state(self):
+        """Current state of the camera."""
+        return self._state
+
+    async def update_image(self, image, filename):
+        """Update the camera image."""
+        if self._state == STATE_IDLE:
+            self._last_trip = dt_util.utcnow()
```

Comment:
It is not, this will refer to the moment we start recording, last_changed will be updated during recording. The duration of the recording will be `last_changed - last_trip`

Reply:
If state is the same, ie recording, last_changed will not be updated after a state update.

Reply:
last_trip is only set on the transition STATE_IDLE -> STATE_RECORDING
last_changed is set on both transitions

both attributes will be equal only for the very first image of an event.